File: python/99_largest_exponential.py
from math import log

def largest_exponential(base_exps: list) -> int:
    max_idx: int = 0
    max_res: int = 0
    
    for i in range(len(base_exps)):
        
        # b^y = x
        # log(b^y) = log(x)
        # y * log(b) = log(x)
        base_exp = base_exps[i] # [base, exponent]
        res = int(base_exp[1]) * log(int(base_exp[0]))
        
        # update 1-based line index with largest found y * log(b)
        if res > max_res:
            max_res, max_idx = res, i + 1
    
    return max_idx

if __name__ == "__main__":
    f = open("99_base_exp.txt", 'r')
    base_exps = [base_exp.strip().split(",") for base_exp in f.readlines()]
    f.close()

    print(largest_exponential(base_exps))

File: cpp/99_largest_exponential.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <cmath>

int largest_exponential(std::vector<std::pair<int, int>> pairs){
    // store maximum value seen and corresponding line number (1-indexed)
    int res = 0;
    double max_val = 0;
    for (int i = 0; i < pairs.size(); i++){
        // extract base and exponent from pair
        int base = pairs[i].first, power = pairs[i].second;
        // calculate exponent * log(base)
        // comparing logarithms of base-exponent pairs
        // uses smaller values while still maintaining ordering
        double val = power * std::log10(base);
        // update maximum value and line number
        if (val > max_val){
            max_val = val;
            res = i + 1;
        }
    }
    return res;
}

int main(){
    // store base-exponent pairs in ordered vector
    std::vector<std::pair<int, int>> pairs;
    std::ifstream file("/uploads/99_base_exp.txt");
    
    std::string line;
    while (std::getline(file, line)) {
        // extract base and exponent from string of 2 comma-separated values
        std::istringstream stream(line);
        std::string base_str, power_str;

        std::getline(stream, base_str, ',');
        std::getline(stream, power_str, ',');

        int base = std::stoi(base_str);
        int power = std::stoi(power_str);

        pairs.emplace_back(base, power);
    }
    file.close();
    
    std::cout << largest_exponential(pairs) << "\n";
}
